src/main.cpp

```diff
@@ -4719,6 +4719,21 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     }
 
 
+    if (!(nLocalServices & NODE_BLOOM) &&
+              (strCommand == "filterload" ||
+               strCommand == "filteradd" ||
+               strCommand == "filterclear"))
+    {
+        if (pfrom->nVersion >= NO_BLOOM_VERSION) {
+            Misbehaving(pfrom->GetId(), 100);
+            return false;
+        }
+        //TODO: Enable this after reasonable network upgrade
+        //else {
+        //    pfrom->fDisconnect = true;
+        //    return false;
+        //}
+    }
```

Comment:
This commit was probably meant to be a no-op refactoring, but there is a subtle difference in behaviour as a result of moving this code before this check:
```
    else if (pfrom->nVersion == 0)
    {
        // Must have a version message before anything else
        Misbehaving(pfrom->GetId(), 1);
        return false;
    }
```
and this one:
```
    // Disconnect existing peer connection when:
    // 1. The version message has been received
    // 2. Overwinter is active
    // 3. Peer version is pre-Overwinter
    else if (NetworkUpgradeActive(GetHeight(), chainparams.GetConsensus(), Consensus::UPGRADE_OVERWINTER)
            && (pfrom->nVersion < chainparams.GetConsensus().vUpgrades[Consensus::UPGRADE_OVERWINTER].nProtocolVersion))
    {
        LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
        pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
                            strprintf("Version must be %d or greater",
                            chainparams.GetConsensus().vUpgrades[Consensus::UPGRADE_OVERWINTER].nProtocolVersion));
        pfrom->fDisconnect = true;
        return false;
    }
```

In the case where the peer sends a prohibited ``filter*`` command before a ``version`` command, the node will call ``Misbehaving`` with a penalty of 100 instead of 1. This seems harmless, and maybe intended. The default ban threshold is 100.

In the case where the peer sends a prohibited ``filter*`` command and has a pre-Overwinter protocol version after Overwinter activation, the node will call ``Misbehaving`` with a penalty of 100, instead of logging, sending a reject message, and disconnecting. If the ban threshold (``-banscore``) is the default 100 then it will be banned anyway; if the ban threshold has been increased then I think the node will still send "reject" and disconnect on the next command that is not ``filter*`` or ``verack``. If ``-enforcenodebloom`` is set then it will be disconnected anyway. This seems acceptable; I just wanted to make sure it is considered.

Reply:
Good spot @daira

Reply:
The handshake requires that the version message must be sent first.  Including the block of code before the version check implies otherwise.

For consistency and clarity, let's make the block of code an 'else if ...' and move it where the filter commands are currently processed, i.e. above `    else if (strCommand == "filterload")`

Reply:
Upstream PR https://github.com/bitcoin/bitcoin/pull/7087 explicitly moved it to this location. Per the commit message, the reason for the move was:

```
Move bloom filter filtering logic outside of command "switch" (giant if/else).

Moving this logic outside of the "switch" makes it far simpler to
enable the forced disconnect by a parameter.
```

I'm fine moving it back, as long as the forced disconnection parameter actually still works.

Reply:
I don't see why the forced disconnect wouldn't work.  I tested this locally and the qa test which uses the -enforcenodebloom parameter passes fine.  https://gist.github.com/bitcartel/4ebeb71855b0035f6f3540de0639c6b2